internal/v4/api_test.go

```diff
@@ -2048,3 +2073,339 @@ func (s *APISuite) TestMacaroon(c *gc.C) {
+	expectStatus: http.StatusProxyAuthRequired,
+	expectBody:   dischargeRequiredBody("http://0.1.2.3/"),
+	expectEntities: []*mongodoc.Entity{
+		storetesting.NewEntity("~charmers/trusty/wordpress-0").WithPromulgatedURL("trusty/wordpress-0").Build(),
+	},
+	expectBaseEntities: []*mongodoc.BaseEntity{
+		storetesting.NewBaseEntity("~charmers/wordpress").WithPromulgated(true).Build(),
+	},
+}, {
+	about: "promulgate base entity with unauthorized user macaroon",
+	entities: []*mongodoc.Entity{
+		storetesting.NewEntity("~charmers/trusty/wordpress-0").Build(),
+	},
+	baseEntities: []*mongodoc.BaseEntity{
+		storetesting.NewBaseEntity("~charmers/wordpress").Build(),
+	},
+	id:   "~charmers/wordpress",
+	body: storetesting.JSONReader(params.PromulgateRequest{Promulgate: true}),
+	caveats: []checkers.Caveat{
+		checkers.DeclaredCaveat(v4.UsernameAttr, "bob"),
+		checkers.DeclaredCaveat(v4.GroupsAttr, "yellow"),
+	},
+	expectStatus: http.StatusUnauthorized,
+	expectBody: params.Error{
+		Message: `unauthorized: access denied for user "bob"`,
+		Code:    params.ErrUnauthorized,
+	},
+	expectEntities: []*mongodoc.Entity{
+		storetesting.NewEntity("~charmers/trusty/wordpress-0").Build(),
+	},
+	expectBaseEntities: []*mongodoc.BaseEntity{
+		storetesting.NewBaseEntity("~charmers/wordpress").Build(),
+	},
+}}
+
+func (s *APISuite) TestPromulgate(c *gc.C) {
+	srv, store := newServer(c, s.Session, nil, charmstore.ServerParams{
+		AuthUsername:     serverParams.AuthUsername,
+		AuthPassword:     serverParams.AuthPassword,
+		IdentityLocation: "http://0.1.2.3/",
+		PublicKeyLocator: bakery.PublicKeyLocatorMap{
+			"http://0.1.2.3/": &testPublicKey,
+		},
+	})
+	for i, test := range promulgateTests {
+		c.Logf("%d. %s\n", i, test.about)
+		_, err := store.DB.Entities().RemoveAll(nil)
+		c.Assert(err, gc.IsNil)
+		_, err = store.DB.BaseEntities().RemoveAll(nil)
+		c.Assert(err, gc.IsNil)
+		for _, e := range test.entities {
+			err := store.DB.Entities().Insert(e)
+			c.Assert(err, gc.IsNil)
+		}
+		for _, e := range test.baseEntities {
+			err := store.DB.BaseEntities().Insert(e)
+			c.Assert(err, gc.IsNil)
+		}
+		if test.method == "" {
+			test.method = "PUT"
+		}
+		header := http.Header{"Content-Type": {"application/json"}}
+		if len(test.caveats) > 0 {
+			m, err := store.Bakery.NewMacaroon("", nil, test.caveats)
+			c.Assert(err, gc.IsNil)
+			cookie, err := httpbakery.NewCookie(macaroon.Slice{m})
+			c.Assert(err, gc.IsNil)
+			header.Set("Cookie", (&http.Cookie{
+				Name:  cookie.Name,
+				Value: cookie.Value,
+			}).String())
+		}
+		httptesting.AssertJSONCall(c, httptesting.JSONCallParams{
+			Handler:      srv,
+			URL:          storeURL(test.id + "/promulgate"),
+			Method:       test.method,
+			Body:         test.body,
+			Header:       header,
+			Username:     test.username,
+			Password:     test.password,
+			ExpectStatus: test.expectStatus,
+			ExpectBody:   test.expectBody,
```

Comment:
This uses enough fields from it that it seems like it might be worth just using JSONCallParams in the test. Something to consider for a future PR.